xwiki-platform-core/xwiki-platform-refactoring/xwiki-platform-refactoring-default/src/main/java/org/xwiki/refactoring/job/XClassDeletingListener.java

```diff
@@ -0,0 +1,172 @@
+        }
+
+        // Ask a confirmation to the user if some pages contain used XClass
+        if (!question.getImpactedObjects().isEmpty()) {
+
+            try {
+                if (this.documentAccessBridge.isAdvancedUser(job.getRequest().getProperty("user.reference"))) {
+                    // Conservative choice: we let the user enable the pages to delete.
+                    question.unselectAll();
+                    // The user can modify the question so it could disable some EntitySelection.
+                    // We add a timeout because when a refactoring job is running, it prevents others to run.
+                    // 5 minutes is probably enough for the user to decide if the process should go on.
+                    boolean ack = job.getStatus().ask(question, 5, TimeUnit.MINUTES);
+                    if (!ack) {
+                        // Without any confirmation, we must cancel the operation.
+                        String message = "The question has been asked, however no answer has been received.";
+                        this.logger.warn(message);
+                        cancelableEvent.cancel(message);
+                    }
+                } else {
+                    question.setRefactoringForbidden(true);
+                    // we don't want the user to answer the question,
+                    // but we want to display that his action is forbidden.
+                    boolean ack = job.getStatus().ask(question, 1, TimeUnit.MINUTES);
+                    if (!ack) {
+                        String message = "The question has been canceled because this refactoring is forbidden.";
+                        cancelableEvent.cancel(message);
+                    }
+                }
+            } catch (InterruptedException e) {
+                this.logger.warn("Confirm question has been interrupted.");
+                cancelableEvent.cancel("Question has been interrupted.");
+            }
+        }
+    }
+
+    private void checkIfDeleteIsAllowed(EntitySelection entitySelection, XClassBreakingQuestion question)
+    {
+        DocumentReference documentReference = (DocumentReference) entitySelection.getEntityReference();
+        String query = "select distinct doc.fullName from XWikiDocument doc, BaseObject obj where "
+            + "doc.fullName=obj.name and obj.className=:className";
+
+        String className = serializer.serialize(documentReference);
+        try {
+            List<String> results = this.queryManager.createQuery(query, Query.HQL)
+                .bindValue("className", className)
+                .setWiki(documentReference.getWikiReference().getName())
+                .<String>execute();
+
+            if (results.isEmpty()) {
+                question.markAsFreePage(entitySelection);
+            } else {
+                for (String documentObjectName : results) {
+                    EntityReference documentObjectReference = this.resolver.resolve(documentObjectName,
```

Comment:
You should pass `documentReference` as 3rd parameter to resolve the reference based on the class instead of the current context which may have nothing to do with the class (it's always main wiki in the context pf the refactoring jobs AFAIK). Also that way you can use the default resolver.

Reply:
ok, thanks for the tip

Reply:
It's a good reflex to always try to find a reference to pass before relying on "current" resolver. It avoid mistakes like this.